### ISSUE TITLE
providedCompile and providedRuntime dependencies do not work as advertised for WAR task

### ISSUE THREAD

**Issue by chrisbrookes** [grails/grails-core#10728]
### Steps to Reproduce

1. create empty app
2. In build.gradle, change a compile or runtime dependency to providedCompile or providedRuntime, E.g.:
`runtime "com.h2database:h2"`
to
`providedRuntime "com.h2database:h2"`

3. run `grails war`

### Expected Behaviour

The build .war file should not contain the jars specified as providedCompile or providedRuntime, nor their transitive dependencies. This is as per the information specified in the [war command line docs](http://docs.grails.org/3.2.11/ref/Command%20Line/war.html):

> Grails 2.x war command allowed you to supply the argument nojars - which packaged the WAR with no jar files. In Grails 3.x, this argument is no longer available. Instead, you may use Gradle capabilities to generate a WAR without jar files. One option would be to set your dependencies as providedCompile and providedRuntime. Those two configurations have the same scope as the respective compile and runtime configurations, except that they are not added to the WAR archive. Check Gradle WAR Plugin documentation for more information.

### Actual Behaviour

The .war file does include the jars in `WEB-INF/lib/`. In the above case the `h2-1.4.194.jar` is still being included.

Note that using `provided` does seem to work - the libs are excluded in the war WEB-INF/lib.

### Environment Information

- **Operating System**: Ubuntu Linux 16.04
- **Grails Version:** 3.2.11
- **JDK Version:** Java(TM) SE Runtime Environment (build 1.8.0_91-b14)

**Comment by jameskleeh** [comment]
Does `./gradlew -Dgrails.env=prod war` produce the desired results?

`grails war` calls `./gradlew assemble`, which has functionality in it for runnable war/jar files

**Comment by chrisbrookes** [comment]
No, it doesn't seem to make a difference using `grails war` or `./gradlew war`.

**Comment by vijaysl** [comment]
I am facing something similar. If I specify 
providedRuntime "org.springframework.boot:spring-boot-starter-tomcat", the tomcat jars are not being excluded from the packaged war. I end up having to use something like below
    if (!project.hasProperty('grailsEnv') || project.grailsEnv.equals('dev') || project.grailsEnv.equals('')) {
        compile "org.springframework.boot:spring-boot-starter-tomcat"
    } else {
        provided "org.springframework.boot:spring-boot-starter-tomcat"
    }

Wondering why the providedRuntime scope from the war plugin is not working

**Comment by graemerocher** [comment]
This is not a bug as such. The reason for this is:

https://github.com/grails/grails-core/blob/master/grails-gradle-plugin/src/main/groovy/org/grails/gradle/plugin/core/GrailsGradlePlugin.groovy#L279

Since `providedRuntime` and `providedCompile` depend on the presence of the Gradle `war` plugin Grails' gradle plugin includes its the `nebula` provided plugins which include a version of `provided` that is not tied to the `war` plugin. The name of this scope is `provided` and Grails overrides the scope name that Spring Boot uses when packaging dependencies.

In other words just using `provided` instead of `providedCompile` or `providedRuntime` it resolves the issue.

**Comment by abcfy2** [comment]
Hi @graemerocher 

I've tried `provided` but it's not working, here is my `build.gradle`

```
buildscript {
    repositories {
        mavenLocal()
        maven { url "https://repo.grails.org/grails/core" }
    }
    dependencies {
        classpath "org.grails:grails-gradle-plugin:$grailsVersion"
        classpath "gradle.plugin.com.energizedwork:idea-gradle-plugins:1.4"
        classpath "org.grails.plugins:hibernate5:${gormVersion - ".RELEASE"}"
        classpath "org.grails.plugins:views-gradle:1.2.6"
    }
}

version "0.1"
group "kaishan.restful"

apply plugin: "eclipse"
apply plugin: "idea"
apply plugin: "war"
apply plugin: "org.grails.grails-web"
apply plugin: "com.energizedwork.idea-project-components"
apply plugin: "org.grails.plugins.views-json"

repositories {
    mavenLocal()
    maven { url "https://repo.grails.org/grails/core" }
}

def esVersion = '5.4.3'
ext['elasticsearch.version'] = esVersion

dependencies {
    compile "org.springframework.boot:spring-boot-starter-logging"
    compile "org.springframework.boot:spring-boot-autoconfigure"
    compile "org.grails:grails-core"
    compile "org.springframework.boot:spring-boot-starter-actuator"
    compile "org.springframework.boot:spring-boot-starter-undertow"
    compile "org.grails:grails-plugin-url-mappings"
    compile "org.grails:grails-plugin-rest"
    compile "org.grails:grails-plugin-codecs"
    compile "org.grails:grails-plugin-interceptors"
    compile "org.grails:grails-plugin-services"
    compile "org.grails:grails-plugin-datasource"
    compile "org.grails:grails-plugin-databinding"
    compile "org.grails:grails-web-boot"
    compile "org.grails:grails-logging"
    compile "org.grails.plugins:cache"
    compile "org.grails.plugins:async"
    compile "org.grails.plugins:events"
    compile "org.grails.plugins:hibernate5"
    compile "org.hibernate:hibernate-core:5.1.5.Final"
    compile "org.grails.plugins:views-json"
    compile "org.grails.plugins:views-json-templates"
    compile 'org.grails.plugins:elasticsearch:2.4.0'
    provided 'org.elasticsearch.plugin:mapper-attachments:2.4.6'
    console "org.grails:grails-console"
    profile "org.grails.profiles:rest-api"
    runtime "mysql:mysql-connector-java:5.1.35"
    testCompile "org.grails:grails-gorm-testing-support"
    testCompile "org.grails:grails-web-testing-support"
}

bootRun {
    jvmArgs('-Dspring.output.ansi.enabled=always')
    addResources = true
}
```

I'm using grails `3.3.2` and I want `org.elasticsearch.plugin:mapper-attachments:2.4.6` not packaging to the war file, but `provided` seem not working.

```
./grailsw war
BUILD SUCCESSFUL in 13s
7 actionable tasks: 6 executed, 1 up-to-date
| Built application to build/libs using environment: production

jar -tf build/libs/ks-rest-0.1.war | grep mapper
WEB-INF/lib-provided/mapper-attachments-2.4.6.jar
```

This dependency locate in `WEB-INF/lib-provided/`, how to exclude totally from war file, but working on `run-app` and `test-app` ? Thanks.
